### PR TITLE
Add wide and deep model into sparse example

### PR DESCRIPTION
**Pull request by ZiyueHuang**
Implement [wide & deep model](https://arxiv.org/abs/1606.07792). Some tricks of feature engineering are adapted from tensorflow's [wide and deep tutorial](https://www.tensorflow.org/tutorials/wide_and_deep). Adult dataset in uci-machine-learning is used.

Running on the machine of 20 cores CPU,
```
python wide_deep_classification.py
```

```
2017-12-07 02:13:11,824 Training started ...
2017-12-07 02:13:14,561 Epoch[0] Batch [100]	Speed: 3687.98 samples/sec	accuracy=0.786100
2017-12-07 02:13:15,801 Epoch[0] Batch [200]	Speed: 8069.39 samples/sec	accuracy=0.796400
2017-12-07 02:13:17,456 Epoch[0] Batch [300]	Speed: 6044.32 samples/sec	accuracy=0.800500
2017-12-07 02:13:19,418 epoch 0, accuracy = 0.817777777778
2017-12-07 02:13:19,428 Saved checkpoint to "checkpoint-0000.params"
2017-12-07 02:13:19,439 Saved optimizer state to "checkpoint-0000.states"
2017-12-07 02:13:20,682 Epoch[1] Batch [100]	Speed: 8195.44 samples/sec	accuracy=0.810000
2017-12-07 02:13:21,522 Epoch[1] Batch [200]	Speed: 11956.14 samples/sec	accuracy=0.827100
2017-12-07 02:13:23,201 Epoch[1] Batch [300]	Speed: 5960.21 samples/sec	accuracy=0.806400
2017-12-07 02:13:25,225 epoch 1, accuracy = 0.831172839506
2017-12-07 02:13:25,240 Saved checkpoint to "checkpoint-0001.params"
2017-12-07 02:13:25,249 Saved optimizer state to "checkpoint-0001.states"
2017-12-07 02:13:27,382 Epoch[2] Batch [100]	Speed: 4703.73 samples/sec	accuracy=0.829000
2017-12-07 02:13:29,687 Epoch[2] Batch [200]	Speed: 4347.43 samples/sec	accuracy=0.835000
2017-12-07 02:13:31,266 Epoch[2] Batch [300]	Speed: 6334.48 samples/sec	accuracy=0.826500
2017-12-07 02:13:32,177 epoch 2, accuracy = 0.841790123457
2017-12-07 02:13:32,192 Saved checkpoint to "checkpoint-0002.params"
2017-12-07 02:13:32,200 Saved optimizer state to "checkpoint-0002.states"
2017-12-07 02:13:34,267 Epoch[3] Batch [100]	Speed: 4844.49 samples/sec	accuracy=0.836600
2017-12-07 02:13:35,971 Epoch[3] Batch [200]	Speed: 5868.73 samples/sec	accuracy=0.842100
2017-12-07 02:13:37,122 Epoch[3] Batch [300]	Speed: 8695.22 samples/sec	accuracy=0.839000
2017-12-07 02:13:39,280 epoch 3, accuracy = 0.843888888889
2017-12-07 02:13:39,298 Saved checkpoint to "checkpoint-0003.params"
2017-12-07 02:13:39,305 Saved optimizer state to "checkpoint-0003.states"
2017-12-07 02:13:41,554 Epoch[4] Batch [100]	Speed: 4475.01 samples/sec	accuracy=0.839800
2017-12-07 02:13:42,552 Epoch[4] Batch [200]	Speed: 10023.14 samples/sec	accuracy=0.843900
2017-12-07 02:13:43,296 Epoch[4] Batch [300]	Speed: 13454.17 samples/sec	accuracy=0.842700
2017-12-07 02:13:45,618 epoch 4, accuracy = 0.84537037037
2017-12-07 02:13:45,626 Saved checkpoint to "checkpoint-0004.params"
2017-12-07 02:13:45,632 Saved optimizer state to "checkpoint-0004.states"
2017-12-07 02:13:47,358 Epoch[5] Batch [100]	Speed: 5830.27 samples/sec	accuracy=0.842600
2017-12-07 02:13:48,330 Epoch[5] Batch [200]	Speed: 10300.89 samples/sec	accuracy=0.845300
2017-12-07 02:13:50,116 Epoch[5] Batch [300]	Speed: 5599.59 samples/sec	accuracy=0.845700
2017-12-07 02:13:51,727 epoch 5, accuracy = 0.847037037037
2017-12-07 02:13:51,741 Saved checkpoint to "checkpoint-0005.params"
2017-12-07 02:13:51,755 Saved optimizer state to "checkpoint-0005.states"
2017-12-07 02:13:52,921 Epoch[6] Batch [100]	Speed: 8609.80 samples/sec	accuracy=0.845900
2017-12-07 02:13:54,538 Epoch[6] Batch [200]	Speed: 6183.48 samples/sec	accuracy=0.842300
2017-12-07 02:13:56,474 Epoch[6] Batch [300]	Speed: 5165.55 samples/sec	accuracy=0.849900
2017-12-07 02:13:57,639 epoch 6, accuracy = 0.846975308642
2017-12-07 02:13:57,655 Saved checkpoint to "checkpoint-0006.params"
2017-12-07 02:13:57,662 Saved optimizer state to "checkpoint-0006.states"
2017-12-07 02:13:59,894 Epoch[7] Batch [100]	Speed: 4534.24 samples/sec	accuracy=0.846700
2017-12-07 02:14:02,040 Epoch[7] Batch [200]	Speed: 4659.63 samples/sec	accuracy=0.842900
2017-12-07 02:14:03,891 Epoch[7] Batch [300]	Speed: 5403.89 samples/sec	accuracy=0.849700
2017-12-07 02:14:05,084 epoch 7, accuracy = 0.85012345679
2017-12-07 02:14:05,102 Saved checkpoint to "checkpoint-0007.params"
2017-12-07 02:14:05,109 Saved optimizer state to "checkpoint-0007.states"
2017-12-07 02:14:06,697 Epoch[8] Batch [100]	Speed: 6467.88 samples/sec	accuracy=0.846600
2017-12-07 02:14:08,491 Epoch[8] Batch [200]	Speed: 5575.99 samples/sec	accuracy=0.843800
2017-12-07 02:14:09,341 Epoch[8] Batch [300]	Speed: 11763.02 samples/sec	accuracy=0.848800
2017-12-07 02:14:10,351 epoch 8, accuracy = 0.850617283951
2017-12-07 02:14:10,370 Saved checkpoint to "checkpoint-0008.params"
2017-12-07 02:14:10,377 Saved optimizer state to "checkpoint-0008.states"
2017-12-07 02:14:11,921 Epoch[9] Batch [100]	Speed: 6502.14 samples/sec	accuracy=0.849600
2017-12-07 02:14:14,121 Epoch[9] Batch [200]	Speed: 4545.56 samples/sec	accuracy=0.845900
2017-12-07 02:14:15,602 Epoch[9] Batch [300]	Speed: 6755.95 samples/sec	accuracy=0.853500
2017-12-07 02:14:16,940 epoch 9, accuracy = 0.851419753086
2017-12-07 02:14:16,946 Saved checkpoint to "checkpoint-0009.params"
2017-12-07 02:14:16,954 Saved optimizer state to "checkpoint-0009.states"
2017-12-07 02:14:16,954 Training completed.
```

cc @eric-haibin-lin  @szha

### CODE REVIEW

File: example/sparse/linear_classification.py [hunk 1]
```diff
@@ -56,19 +81,34 @@
...
     num_worker = kv.num_workers if kv else 1
 
     # dataset
-    num_features = AVAZU['num_features']
-    data_dir = os.path.join(os.getcwd(), 'data')
-    train_data = os.path.join(data_dir, AVAZU['train'])
-    val_data = os.path.join(data_dir, AVAZU['test'])
-    get_libsvm_data(data_dir, AVAZU['train'], AVAZU['url'])
-    get_libsvm_data(data_dir, AVAZU['test'], AVAZU['url'])
+    if args.model == 'linear':
```
Comment (szha) [inline example/sparse/linear_classification.py]: Why is dataset dependent on the type of model?

Comment (ZiyueHuang) [reply example/sparse/linear_classification.py]: Other implementations (Keras, Tensorflow, etc.) use adult dataset and I think it's better to be comparable here. And feature engineering is important. Avazu libsvm dataset cannot be directly used in wide and deep model because of some feature engineering issues, for example, there is one format issue, features in one side (wide or deep) should be together.

File: example/sparse/linear_classification.py [hunk 2]
```diff
@@ -56,19 +81,34 @@
...
+        train_data = os.path.join(data_dir, AVAZU['train'])
+        val_data = os.path.join(data_dir, AVAZU['test'])
+        get_libsvm_data(data_dir, AVAZU['train'], AVAZU['url'])
+        get_libsvm_data(data_dir, AVAZU['test'], AVAZU['url'])
+        # The positive class weight, says how much more we should upweight the importance of
+        # positive instances in the objective function.
+        # This is used to combat the extreme class imbalance.
+        model = linear_model(AVAZU['num_features'], AVAZU['positive_class_weight'])
+    else:
+        num_features = ADULT['num_features']
```
Comment (eric-haibin-lin) [inline example/sparse/linear_classification.py]: I think this example should be moved to another file since it's not a linear model anymore.

File: example/sparse/wide_deep_model.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import mxnet as mx
+from weighted_softmax_ce import *
+
+
+def wide_deep_model(num_linear_features, num_embed_features, num_cont_features, input_dims, hidden_units, positive_cls_weight):
```
Comment (eric-haibin-lin) [inline example/sparse/wide_deep_model.py]: This line is too long and hurts readability. Could you split it into multiple lines?
@szha Did we disable lint in example folders?

File: example/sparse/wide_deep_model.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    data = mx.symbol.Variable("data", stype='csr')
+    label = mx.symbol.Variable("softmax_label")
+
+    x = mx.symbol.slice_axis(data=data, axis=1, begin=0, end=num_linear_features)
+    x = mx.symbol.cast_storage(x, 'csr')
```
Comment (eric-haibin-lin) [inline example/sparse/wide_deep_model.py]: I think we should implement `slice_axis` for `csr` first. I don't think it's a good idea to add `cast_storage` in the network in the example.

File: example/sparse/linear_classification.py (hunk 2 above)
Comment (ZiyueHuang) [reply example/sparse/linear_classification.py]: linear classification and wide & deep classification are separated to two files now.

File: example/sparse/wide_deep_model.py (hunk 3 above)
Comment (ZiyueHuang) [reply example/sparse/wide_deep_model.py]: Split to two lines.

File: example/sparse/wide_deep_model.py (hunk 4 above)
Comment (ZiyueHuang) [reply example/sparse/wide_deep_model.py]: `cast_storage ` is removed now and this will cause `adam` fallback. I will work on `slice_axis ` for `csr` these days but it will not affect the codes here.

**Comment by ZiyueHuang** [comment]
@eric-haibin-lin row_ids corresponding to embedding or continuous features are removed and I think distributed version should be okay. But my machine seems have some weird problems with protobuf. Could you please check the distributed version?
```
../../tools/launch.py -n 2 --launcher=local python wide_deep_classification.py --kvstore=dist_async
```

**Comment by eric-haibin-lin** [comment]
@ZiyueHuang what problem do you see? What's the error message?
BTW, we should hold this PR off until sparse embedding is ready.

File: example/sparse/wide_deep_model.py (hunk 3 above)
Comment (szha) [reply example/sparse/wide_deep_model.py]: I don't think examples were ever included in lint check, though I think it would be great if it's turned on.

File: example/sparse/wide_deep_classification.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    train_data = os.path.join(data_dir, ADULT['train'])
+    val_data = os.path.join(data_dir, ADULT['test'])
+    train_csr, train_dns, train_label = get_uci_adult(data_dir, ADULT['train'], ADULT['url'])
+    val_csr, val_dns, val_label = get_uci_adult(data_dir, ADULT['test'], ADULT['url'])
+
+    model = wide_deep_model(ADULT['num_linear_features'], ADULT['num_embed_features'], ADULT['num_cont_features'],
+                            ADULT['embed_input_dims'], ADULT['hidden_units'], ADULT['positive_class_weight'])
+
+    # data iterator
+    train_data = mx.io.NDArrayIter({'csr_data': train_csr, 'dns_data': train_dns},
```
Comment (eric-haibin-lin) [inline example/sparse/wide_deep_classification.py]: add a wrapper of train_data = mx.io.PrefetchingIter(train_data) will help hide io latency

File: example/sparse/wide_deep_model.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    hidden = mx.symbol.concat(*features, dim=1)
+    hidden = mx.symbol.BatchNorm(data=hidden)
+    hidden = mx.symbol.FullyConnected(data=hidden, num_hidden=hidden_units[1])
+    hideen = mx.symbol.Activation(data=hidden, act_type='relu')
+    hidden = mx.symbol.FullyConnected(data=hidden, num_hidden=hidden_units[2])
+    hideen = mx.symbol.Activation(data=hidden, act_type='relu')
+    deep_out = mx.symbol.FullyConnected(data=hidden, num_hidden=2)
+
+    out = mx.symbol.Custom(linear_out+deep_out, label, op_type='weighted_softmax_ce_loss',
```
Comment (eric-haibin-lin) [inline example/sparse/wide_deep_model.py]: Does this dataset also have imbalanced label problem? Does the WD model in TF use any weighted ce loss?

Comment (ZiyueHuang) [reply example/sparse/wide_deep_model.py]: Yes, there are 24720 for the first class and 7841 for the second class. TF version doesn't use weighted ce loss.

**Comment by mbaijal** [comment]
@ZiyueHuang 
This PR is failing in a test. Is this caused by the changes here?

**Comment by SSSSQD** [comment]
hi, wide & deep learning, model.py line 52 and line 54, the hideen should be hidden

**Comment by ZiyueHuang** [comment]
@SSSSQD Thanks for pointing this mistake out. I will fix it now.
